### ISSUE TITLE
cuda implementation for fp16

### ISSUE THREAD

**Issue by yaseryacoob** [mhamilton723/FeatUp#36]
I tried to convert your code to fp16 to see how the upsampling looks,. It turns out that the cuda part you implemented doesn't handle fp16. (I got this error):
```
File "~/FeatUp/featup/adaptive_conv_cuda/adaptive_conv.py", line 21, in forward
    result = cuda_impl.forward(input, filters)
RuntimeError: "adaptive_conv_forward_cuda" not implemented for 'Half'
```

I think it is worth supporting fp16 (and my naive guess it is not too much trouble).

What's your take on this?

many thanks

**Comment by mhamilton723** [comment]
Just merged the fix for this, thanks for finding the issue

**Comment by mhamilton723** [comment]
Sorry for accidentally closing this, our CUDA guru @axelfeldmann  will be able to take a look at this on april 19th, but in the meantime if you find a fix we would gladly accept the PR.

**Comment by yaseryacoob** [comment]
Thanks Mark, I was sorting through your latest push and didn't see anything relevant. I will wait for @axelfeldman, I just don't know enough to make the changes. I observed another issue, will raise in a new issue.

**Comment by mhamilton723** [comment]
@yaseryacoob yes that was to fix multi-gpu environments. It seems fp16 still needs some other fixes

**Comment by mhamilton723** [comment]
@yaseryacoob , @axelfeldmann just contributed the fix for this!

**Comment by yaseryacoob** [comment]
I am not sure I am gettingit  right. Looked at the repo and downloaded the new adaptive_conv_kernel.cu (which appears to have the following changes 

```diff adaptive_conv_kernel.cu adaptive_conv_kernel_pre_fp16.cu
4d3
< #include <ATen/cuda/CUDAContext.h>
10d8
< using namespace at;
129d126
<     at::cuda::set_device(input.device().index());
163c160
<         AT_DISPATCH_FLOATING_TYPES_AND_HALF(out.scalar_type(), "adaptive_conv_forward_cuda", ([&] {
---
>         AT_DISPATCH_FLOATING_TYPES(out.scalar_type(), "adaptive_conv_forward_cuda", ([&] {
180d176
<     at::cuda::set_device(grad_output.device().index());
216c212
<         AT_DISPATCH_FLOATING_TYPES_AND_HALF(out.scalar_type(), "adaptive_conv_grad_input_cuda", ([&] {
---
>         AT_DISPATCH_FLOATING_TYPES(out.scalar_type(), "adaptive_conv_grad_input_cuda", ([&] {
232d227
<     at::cuda::set_device(grad_output.device().index());
271c266
<         AT_DISPATCH_FLOATING_TYPES_AND_HALF(out.scalar_type(), "adaptive_conv_grad_filters_cuda", ([&] {
---
>         AT_DISPATCH_FLOATING_TYPES(out.scalar_type(), "adaptive_conv_grad_filters_cuda", ([&] {
```   

 I also removed the  __pycache__ so that new .pyc are created. 

The error is the same I had first time around (see initial issue).   

Can you please provide a full example, in case I missing on something else?   Thanks!

**Comment by axelfeldmann** [comment]
With my install, this works:

```
import torch
from featup.adaptive_conv_cuda.adaptive_conv import AdaptiveConv

a = torch.randn((1, 3, 10, 10), dtype=torch.float16).cuda()
b = torch.randn((1, 8, 8, 3, 3), dtype=torch.float16).cuda()

result = AdaptiveConv.apply(a, b)
assert result.dtype == torch.float16
```

I would recommend the following:
```
pip uninstall FeatUp
rm -r build # the build directory inside of your local FeatUp repo
pip install .
```
and let me know if that fixes it.

**Comment by yaseryacoob** [comment]
Thanks @axelfeldmann I can confirm everything worked fine, there was a adaptive_conv_cpp_impl.cpython-310-x86_64-linux-gnu.so in the FeatUp root that was messing things up, once removed got fixed.

My simple test of fp32 versus fp16 on the plants.png showed 0.991 cosine similarity between the features (mean for all upsampled features). Visually I couldn't see differences. The advantage is that larger images can be processed on a given GPU. The issue of waviness I reported in #39 remains a challenging issue
![YYdinov2](https://github.com/mhamilton723/FeatUp/assets/8571131/94af3760-dfd3-4c44-be38-635ccf26fbd7)
